### DISCUSSION TITLE
Automatically Configure DB in Docker Container

### DISCUSSION THREAD

**Discussion by casrar** [pocketbase/pocketbase#4289] [Q&A] [answered]
I want to figure out a way that on startup of a docker container of PB, I can either configure a DB from scratch or utilize an existing pb_data dir. I've been really scratching my head over this so any advice on a idiomatic or just a not horrendous way to do this would be appreciated. During my research I've come across these threads/posts, [Create first admin from code #2933](https://github.com/pocketbase/pocketbase/discussions/2933), [Docker automatically setup #1051](https://github.com/pocketbase/pocketbase/discussions/1051), [Going to production](https://pocketbase.io/docs/going-to-production/). 

Originally I was planning on using python or bash to try and do some of this config but I also thought maybe using Go and interacting with the DB through the SDK would be easier. Any ideas/help is appreciated, or even more resources to look at.

tldr; I need a way to automatically configure Admin and DB tables and schema, or use the existing pd_data dir.

**Comment by ganigeorgiev** [comment]
You could check the [Go](https://pocketbase.io/docs/go-migrations/) or [JS](https://pocketbase.io/docs/js-migrations/) migrations guide.

If in your Docker container you are using the prebuilt executable then you can copy the local `pb_migrations` dir into the image (or mount it as a volume) and place it next to your `pb_data`. All new/unapplied migrations will be automatically executed on `serve`.

**Reply by casrar** [reply]
What all does the migration file cover? Is it just the collection schema?

**Reply by ganigeorgiev** [reply]
> ... DB structure, create collections programmatically, initialize default settings and/or run anything that needs to be executed only once.

See also the [Examples](https://pocketbase.io/docs/js-migrations/#examples) section at the end of the linked pages.

**Reply by casrar** [reply]
Ah, thank you! 

 That covers at least the setup of the collections, but in terms of checking/creating admin does it just make sense to do that with a python/bash script or actually write a small go file which uses the sdk? Appreciate the help!

**Reply by ganigeorgiev** [reply]
You refuse to explore the linked resources so I'm not sure how to help you further.
For "checking/creating admin" I would personally use the migrations. See the examples in:

- Go: https://pocketbase.io/docs/go-migrations/#creating-new-admin
- JS: https://pocketbase.io/docs/js-migrations/#creating-new-admin

**Reply by casrar** [reply]
Not refusing, just getting an opinion. Was planning on diving deeper into the link you sent. Appreciate you linking the specific section 👍

**Reply by casrar** [reply, marked as answer]
Just read through the links you sent more deeply (and with a fresher brain) and it looks like using a [js](https://pocketbase.io/docs/js-migrations/) migrations file and the [JSVM](https://pocketbase.io/jsvm/index.html) enable me to do exactly what I want, thanks!

**Reply by mvllow** [reply]
I'm going through the same situation and am using migrations for creating an initial admin and a few users but was curious if there's a better workflow than the following:

1. Add test users via a migration
2. Change user schema to include `role`, for example
3. Add test users via a migration (this time including a role for each user)
4. Delete the test users migration from step 1

Ideally there would be a way to not have to manually go back and remove any migrations. Happy to start a new discussion if necessary but this seemed pretty similar ♥
